File: trunk/sdk/client/src/serverconn/transporter/transporter.js
/***  HIMLE RIA SYSTEM
  ** 
  **  Copyright (C) 2008 HIMLE GROUP http://himle.sorsacode.com/
  **  Copyright (C) 2006-2007 Helmi Technologies Inc.
  ** 
  **  This program is free software; you can redistribute it and/or modify it under the terms
  **  of the GNU General Public License as published by the Free Software Foundation;
  **  either version 2 of the License, or (at your option) any later version. 
  **  This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
  **  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  **  See the GNU General Public License for more details. 
  **  You should have received a copy of the GNU General Public License along with this program;
  **  if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
  ***/

/** class: HTransporter
  *
  * *Simple mid-level AJAX communication system.*
  *
  * Designed as single instance, depends on <HValueManager>.
  *
  * When implementing the server part:
  *  - feed it with raw javascript
  *  - override ses_id as early as possible with your own session id, this tells the clients apart.
  *  - change the syncDelay depending on how fast you want the client to poll the server (a value in ms)
  *  - you may refer to 't' as the HTransporter's namespace.
  *
  * Sample initialization sequence:
  * > t.ses_id = 'nhHOZ8Zo64Wfo';
  * > t.syncDelay = 400;
  * > t.url_base = '/phase2';
  *
  * See Also:
  *  <HValueManager.toXML>
  **/

/* int: HTransportURL
 *
 * Tells the <HTransporter> which url/path to start polling.
 * - "global"
 * - Override it with your server's path or url before the document is loaded.
 */
HTransportURL = false;

/* int: HTransportPoll
 *
 * Tells the <HTransporter> whether to use polling (true) or to sync values only
 * when there's something on the client-side to sync (false).
 * - "global"
 * - Defaults to true (polling), override it with false before the document is
 *   loaded if you don't want polling.
 */
HTransportPoll = true;

/* vars: Instance variables
 *  url_base  - The URL (or path) that the requests are sent to
 *  ses_id    - A value that is the reported in each request by the key 'ses_id'
 *  syncDelay - An integer value (in ms) that the client waits before starting the next request
 */

HTransporter = Base.extend({
  
  constructor: null,
  
  start: function(_url_base){
    this.url_base  = _url_base;
    this.ses_id    = 0;
    this.err_msg   = '';
    this.isBusy    = false;
    this.syncNum   = 0;
    this.syncDelay = 0;
    this.pollMode  = HTransportPoll;
    this.req_timeout = setTimeout('HTransporter.sync();',this.syncDelay);
  },
  
  setPollMode: function(_flag) {
    this.pollMode = _flag;
  },
  
  failure: function(resp){
    alert('HTTP ERROR! STOP.');
    clearTimeout(this.req_timeout);
    this.isBusy = true;
  },
  
  respond: function(resp){
    var _respText = resp.responseText;
    //console.log( _respText );
    //var t = HTransporter;
    //try {
      //t.err_msg = '';
      eval(_respText); 
    //} catch(e) {
    //  t.err_msg = '&err_msg='+e;
    //}
    HTransporter.isBusy = false;
    //HTransporter.isBusy = true;
  },
  
  sync: function(){
    var valid_delay = ((this.syncDelay&&this.syncDelay>0)||this.syncDelay==0);
    // Negative syncDelay stops transporter.
    if(valid_delay && this.url_base){
      if(!this.isBusy){
        this.isBusy = true;
        var _syncData = HValueManager.toXML();
        
        if ("" != _syncData || this.pollMode) {
          
          this.syncNum++;
          HVM.isGetting=true;
          req_args = {
            onSuccess: function(resp){HTransporter.respond(resp);},
            onFailure: function(resp){HTransporter.failure(resp);},
            method:    'post',
            postBody:  'ses_id='+HTransporter.ses_id+_syncData
          };
          this.req  = new Ajax.Request( this.url_base, req_args );
          HVM.isGetting=false;
        }
        else {
          this.isBusy = false;
        }
        
      }
      this.req_timeout = setTimeout('HTransporter.sync();',this.syncDelay);
    }
  },
  
  stop: function() {
    clearTimeout(this.req_timeout);
  }
  
});

LOAD("HTransporter.start(HTransportURL);");

File: experimental/server/lib/transporter/transporter.rb
###
  # HIMLE RIA Server
  # Copyright (C) 2008 HIMLE GROUP http://himle.sorsacode.com/
  # Copyright (C) 2006-2007 Helmi Technologies Inc.
  #  
  #  This program is free software; you can redistribute it and/or modify it under the terms
  #  of the GNU General Public License as published by the Free Software Foundation;
  #  either version 2 of the License, or (at your option) any later version. 
  #  This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
  #  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  #  See the GNU General Public License for more details. 
  #  You should have received a copy of the GNU General Public License along with this program;
  #  if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
  ###

## Uncomment, when Rack supports chunked transfers
#require 'zlib'
#
#class GZString < String
#  alias write <<
#end

=begin
  Transporter is the counterpart to the client's HTransporter xhr engine.
=end
class Transporter
  
  def initialize
    @config = $config[:transporter_conf]
  end
  
  def shutdown
    PLUGINS.shutdown
    VALUES.shutdown
    SESSION.shutdown
  end
  
  ## 
  def xhr(request, response, cookies=false)
    
    ## The response status should always be 200 (OK)
    response.status = 200
    
    # It's better to evaluate plain text than to respond with js.
    response.content_type = 'text/javascript; charset=utf-8'
    response['Cache-Control'] = 'no-cache'
    
    ## Uncomment, when Rack supports chunked transfers:
    #if request['Accept-Encoding'] and request['Accept-Encoding'].include?('gzip') and not $config[:no_gzip]
    #  response.chunked = true
    #  response['Content-Encoding'] = 'gzip'
    #  do_gzip = true
    #else
    #  do_gzip = false
    #end
    
    msg = SESSION.init_msg( request, response, cookies )
    
    if request.query.has_key?('err_msg')
      if DEBUG_MODE
        puts
        puts "CLIENT ERROR:"
        pp request.query['err_msg']
        puts
      end
      
      SESSION.stop_client_with_message( msg,
        @config[:messages][:client_error][:title],
        @config[:messages][:client_error][:descr]+request.query['err_msg'].inspect,
        @config[:messages][:client_error][:uri]
      )
    end
    
    #puts "----- valid session: #{msg.inspect} -----"
    
    if msg.ses_valid
      
      if cookies
        msg.reply('HTransporter.url_base="/x";')
      end
      if (msg.new_session or msg.restored_session) and DEBUG_MODE
        puts
        puts "new session. rescanning apps."
        puts
        FILECACHE.check_scan
        PLUGINS.rescan()
        puts
        puts "rescan done"
        puts
      end
      
      ## Pass the client XML to VALUES
      if request.query.has_key?( 'HSyncData' )
        syncdata_str = request.query[ 'HSyncData' ]
        VALUES.xhr( msg, syncdata_str )
      end
      
      VALUES.validate( msg )
      
      if msg.restored_session
        msg.session[:deps] = []
        PLUGINS.delegate( 'restore_ses', msg )
      elsif msg.new_session
        PLUGINS.delegate( 'init_ses', msg )
      end
      
      ### Allows every application to respond to the idle call
      PLUGINS.idle( msg )
      
      ### Process outgoing values to client
      VALUES.sync_client( msg )
      
    end
    
    ## return the output
    #if do_gzip
    #  outp = GZString.new('')
    #  gzwriter = Zlib::GzipWriter.new(outp,Zlib::BEST_SPEED)
    #  gzwriter.write( msg.output.join("\r\n") )
    #  gzwriter.close
    #else
    #outp = msg.output.join("\r\n")
    #end
    response['Content-Size'] = response.body.size
    #return outp
    #response.body = outp
  end
  
end

